From 2493fc689778cf99756b39285d4699383e286d9f Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:47 -0400
Subject: [fs] btrfs: fix NULL pointer crash of deleting a seed device

Message-id: <1407179109-13714-341-git-send-email-sandeen@redhat.com>
Patchwork-id: 87292
O-Subject: [RHEL7.1 PATCH 340/422] Btrfs: fix NULL pointer crash of deleting a seed device
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Liu Bo <bo.li.liu@oracle.com>

commit 29cc83f69c8338ff8fd1383c9be263d4bdf52d73
Author: Liu Bo <bo.li.liu@oracle.com>
Date:   Sun May 11 23:14:59 2014 +0800

    Btrfs: fix NULL pointer crash of deleting a seed device

    Same as normal devices, seed devices should be initialized with
    fs_info->dev_root as well, otherwise we'll get a NULL pointer crash.

    Cc: Chris Murphy <lists@colorremedies.com>
    Reported-by: Chris Murphy <lists@colorremedies.com>
    Signed-off-by: Liu Bo <bo.li.liu@oracle.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/volumes.c b/fs/btrfs/volumes.c
index d56b8ca..8f22a29 100644
--- a/fs/btrfs/volumes.c
+++ b/fs/btrfs/volumes.c
@@ -6139,10 +6139,14 @@ void btrfs_init_devices_late(struct btrfs_fs_info *fs_info)
  struct btrfs_fs_devices *fs_devices = fs_info->fs_devices;
  struct btrfs_device *device;
 
- mutex_lock(&fs_devices->device_list_mutex);
- list_for_each_entry(device, &fs_devices->devices, dev_list)
-  device->dev_root = fs_info->dev_root;
- mutex_unlock(&fs_devices->device_list_mutex);
+ while (fs_devices) {
+  mutex_lock(&fs_devices->device_list_mutex);
+  list_for_each_entry(device, &fs_devices->devices, dev_list)
+   device->dev_root = fs_info->dev_root;
+  mutex_unlock(&fs_devices->device_list_mutex);
+
+  fs_devices = fs_devices->seed;
+ }
 }
 
 static void __btrfs_reset_dev_stats(struct btrfs_device *dev)
-- 
1.7.1